### PR TITLE
Rewrite the requirement and marker parser

### LINKED ISSUE #618: `Requirement('installer')` errors after 2bd5da391c302f2f5a18fd3e2bd9fb3b75c02e34
```python
from packaging.requirements import Requirement
Requirement('installer')
```

The regression has been introduced in [2bd5da391c302f2f5a18fd3e2bd9fb3b75c02e34](https://github.com/pypa/packaging/commit/2e5593c1288b9f1b57bb1f79acc0ed5fd1b7009c).

### PR DESCRIPTION
**Pull request by pradyunsg**
Closes #618 

The main changes here are:

- Couple the tokenisation with the parsing, with tokens being generated based on what the parser is expecting.
- Explicitly handle whitespace as a token, explicitly "consuming" it during parsing.
- Better error messages, with spans (partly inspired by Python 3.11's error message implementation).
- Parse markers as a part of the parsing a requirement string, instead of delegating to `Marker`.

The code changes here are ready for review, but I'll keep this as a draft while I work on adding better testing for the parser, including for the edge cases that I've identified over the course of this rework.

/cc @hrnciar

### CODE REVIEW

File: packaging/_parser.py [hunk 1]
```diff
@@ -54,141 +49,228 @@ def serialize(self) -> str:
...
+
+    if tokenizer.check("AT"):
+        tokenizer.read()
+        tokenizer.consume("WS")
+        url_start = tokenizer.position
+        url = tokenizer.expect("URL", expected="URL after @").text
+        if tokenizer.check("END", peek=True):
+            return (url, specifier, marker)
+
+        tokenizer.expect("WS", expected="whitespace after URL")
```
Comment (Jackenmen) [inline packaging/_parser.py:107]: It seems that this resolves #529 (and therefore #592) too, could probably add closing keywords for them in the PR description :)

**Comment by pradyunsg** [comment]
Ahaha, I covered an error case that the existing tests don't exercise triggering coverage-related CI errors.

File: packaging/_parser.py (hunk 1 above)
Comment (pradyunsg) [reply packaging/_parser.py:107]: Indeed!

**Comment by pradyunsg** [comment]
Alrighty, this is ready for review. I think the easiest way to review the changes here would be to look at the test suite; and then to go commit-by-commit. :)

**Comment by pradyunsg** [comment]
"How do you relax after dealing with a tricky Python packaging thing at work?"
"I spend 3 hours polishing up a parser I worked on and then break up a giant WIP commit into something legible."

This PR is now ready for review, with rewritten tests for the requirements parser. I like them because they test a *lot* of test strings (the earlier test suite only checked the parser against ~45 strings in the entire run), many of these strings helped identify logical mistakes in the parser's reworking. Notably, the parser on `main` fails various `-k basic` tests added in this PR, that 21.3 as well as this PR pass.

I reckon the easiest way to review the changes here would be to look at the test suite in its final state; and then to go commit-by-commit to see how things evolve. There's still a giant commit with the entire rewrite -- couldn't figure out how to break that up in the time I had at hand. :)

**Comment by pradyunsg** [comment]
@uranusjr @encukou @hroncok @hrnciar: Y'all reviewed #484, so... would one or more of you be willing to review this PR (which is basically a follow-up to that)? Sorry for the spammy mention. 😅

**Comment by hrnciar** [comment]
I went through all commits and nothing else struck my eye. Looks good to me. Thank you for the improvements :) .

**Comment by pradyunsg** [comment]
I've gone through the issue tracker to check which other issues this PR fixes, and... it's quite a few. I stretched a little bit and also covered #432 in this.

**Review by brettcannon — APPROVED** [review]
Quick perusal didn't turn anything up, but I'm not a parsing expert. 😅

**Comment by pradyunsg** [comment]
Thanks @brettcannon and @hrnciar for the reviews!

**Review by hrnciar — COMMENTED** [review]
Apparently, my comment wasn't posted and it was hanging in the pending state. I'll file PR to remove it.

**Review by cicakkobeng8p — COMMENTED** [review]
Summit
